Core/GDCore/IDE/Events/ExpressionTypeFinder.h

```diff
@@ -0,0 +1,189 @@
+class Expression;
+class ObjectsContainer;
+class Platform;
+class ParameterMetadata;
+class ExpressionMetadata;
+}  // namespace gd
+
+namespace gd {
+
+/**
+ * \brief Find the type of the expression or sub-expression that a given node
+ * represents.
+ *
+ * \see gd::ExpressionParser2
+ */
+class GD_CORE_API ExpressionTypeFinder : public ExpressionParser2NodeWorker {
+ public:
+
+  /**
+   * \brief Helper function to find the type of the expression or
+   * sub-expression that a given node represents.
+   */
+  static const gd::String GetType(const gd::Platform &platform,
+                      const gd::ObjectsContainer &globalObjectsContainer,
+                      const gd::ObjectsContainer &objectsContainer,
+                      const gd::String &rootType,
+                      gd::ExpressionNode& node) {
+    gd::ExpressionTypeFinder typeFinder(
+        platform, globalObjectsContainer, objectsContainer, rootType);
+    node.Visit(typeFinder);
+    return typeFinder.GetType();
+  }
+
+  virtual ~ExpressionTypeFinder(){};
+
+ protected:
+  ExpressionTypeFinder(const gd::Platform &platform_,
+                       const gd::ObjectsContainer &globalObjectsContainer_,
+                       const gd::ObjectsContainer &objectsContainer_,
+                       const gd::String &rootType_)
+      : platform(platform_),
+        globalObjectsContainer(globalObjectsContainer_),
+        objectsContainer(objectsContainer_),
+        rootType(rootType_),
+        type("unknown"),
+        child(nullptr) {};
+
+  const gd::String &GetType() {
+    return ConvertSubtype(type);
+  };
+
+  void OnVisitSubExpressionNode(SubExpressionNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitOperatorNode(OperatorNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitUnaryOperatorNode(UnaryOperatorNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitNumberNode(NumberNode& node) override {
+    type = "number";
+  }
+  void OnVisitTextNode(TextNode& node) override {
+    type = "string";
+  }
+  void OnVisitVariableNode(VariableNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitVariableAccessorNode(VariableAccessorNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitIdentifierNode(IdentifierNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitEmptyNode(EmptyNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitObjectFunctionNameNode(ObjectFunctionNameNode& node) override {
+    VisitParent(node);
+  }
+  void OnVisitVariableBracketAccessorNode(
+      VariableBracketAccessorNode& node) override {
+    if (child == nullptr) {
+      type = "unknown";
+    }
+    auto leftSideType = gd::ExpressionLeftSideTypeFinder::GetType(
+        platform, 
+        globalObjectsContainer,
+        objectsContainer,
+        node);
+    if (leftSideType == "number" || leftSideType == "string") {
+      type = leftSideType;
+    }
+    else {
+      type = "number|string";
```

Comment:
The `type` returned by this worker is a mix of:
- an expected type looking up like a parameter declaration
- an actual type looking down, but only looking at the most left branch (using ExpressionLeftSideTypeFinder)

This logic was built with the constraint of following a parser that can't know the right side. Now that it is extracted, it could be enhanced but we probably want to wait a stable version before doing it.

About this case specifically, it must be understood as follow "In a general manner, a bracket variable assessor expression is either a number or a string. According to the actual used expression, which one is it?". The type is then used to give more precise errors or autocompletions focusing on either number or string.